devito/types/sparse.py

```diff
@@ -931,61 +1035,157 @@ class PrecomputedSparseFunction(AbstractSparseFunction):
+        except AttributeError:
+            return None
 
-    def _dist_scatter(self, data=None):
-        data = data if data is not None else self.data
-        distributor = self.grid.distributor
+    @property
+    def coordinates(self):
+        return self._coordinates
 
-        # If not using MPI, don't waste time
-        if distributor.nprocs == 1:
-            return {self: data, self.gridpoints: self.gridpoints.data,
-                    self._interpolation_coeffs: self._interpolation_coeffs.data}
+    @property
+    def coordinates_data(self):
+        try:
+            return self.coordinates.data._local
+        except AttributeError:
+            return None
 
-        raise NotImplementedError
+    @property
+    def dist_origin(self):
+        return self._dist_origin
 
-    def _dist_gather(self, data):
-        distributor = self.grid.distributor
+    @property
+    def interpolation_coeffs(self):
+        """ The Precomputed interpolation coefficients."""
+        return self._interpolation_coeffs
 
-        # If not using MPI, don't waste time
-        if distributor.nprocs == 1:
-            return
+    @property
+    def interpolation_coeffs_data(self):
+        return self.interpolation_coeffs.data._local
 
-        raise NotImplementedError
+    @cached_property
+    def _coordinate_symbols(self):
+        """Symbol representing the coordinate values in each dimension."""
+        p_dim = self.indices[self._sparse_position]
+        if self.gridpoints is not None:
+            return tuple([self.gridpoints.indexify((p_dim, di)) * d.spacing + o
+                          for ((di, d), o) in zip(enumerate(self.grid.dimensions),
+                                                  self.grid.origin)])
 
-    def _arg_apply(self, *args, **kwargs):
-        distributor = self.grid.distributor
+        return tuple([self.coordinates.indexify((p_dim, i))
+                      for i in range(self.grid.dim)])
 
-        # If not using MPI, don't waste time
-        if distributor.nprocs == 1:
-            return
+    @cached_property
+    def _coordinate_indices(self):
+        """
+        Symbol for each grid index according to the coordinates.
 
-        raise NotImplementedError
+        Notes
+        -----
+        The expression `(coord - origin)/spacing` could also be computed in the
+        mathematically equivalent expanded form `coord/spacing -
+        origin/spacing`. This particular form is problematic when a sparse
+        point is in close proximity of the grid origin, since due to a larger
+        machine precision error it may cause a +-1 error in the computation of
+        the position. We mitigate this problem by computing the positions
+        individually (hence the need for a position map).
+        """
+        if self.gridpoints is not None:
+            ddim = self.gridpoints.dimensions[-1]
+            return tuple([self.gridpoints._subs(ddim, di) for di in range(self.grid.dim)])
```

Comment:
I think I prefer `n` for numbers instead of `di` (or `d`) which is typically used for Dimensions instead

Reply:
I like to use `(namei, name) in zip(...)` so that's why I have `di` since the "name" is `d` but can change to n